source/iopipe/stream.d

```diff
@@ -77,3 +79,114 @@ struct ZeroDev
 
 /// Common instance of ZeroDev to use anywhere needed.
 immutable ZeroDev zeroDev;
+
+// helper for copying data from a src range into a buffer
+private size_t copy(Src, Buf)(ref Src src, Buf buf) if(is(immutable(ElementEncodingType!Buf) == immutable(ElementEncodingType!Src)))
+{
+    // element compatible ranges. Use standard copying
+    static if(is(Buf : T[], T) && is(Src : U[], U) && is(immutable(T) == immutable(U)))
+    {
+        // both dynamic arrays, use slice assign.
+        auto n = src.length < buf.length ? src.length : buf.length;
+        buf[0 .. n] = src[0 .. n];
+        src = src[n .. $];
+        return n;
+    }
+    else
+    {
+        size_t n = 0;
+        while(n < buf.length && !src.empty)
+        {
+            buf[n++] = src.front;
+            src.popFront;
+        }
+        return n;
+    }
+}
+
+struct RangeInput(R) {
+    private import std.traits: isNarrowString;
+    private {
+        R src;
+        static if(is(ElementType!R == T[], T))
+        {
+            enum isRangeOfSlices = true;
+            T[] data;
+        }
+        else
+            enum isRangeOfSlices = false;
+    }
+
+    this(R src)
+    {
+        this.src = src;
+        static if(isRangeOfSlices)
+            if(!this.src.empty)
+                data = this.src.front;
+    }
+
+    size_t read(Buf)(Buf buf) if (isNarrowString!Buf || isRandomAccessRange!Buf)
+    {
+        static if(is(typeof(copy(src, buf))))
+        {
+            return copy(src, buf);
+        }
+        else static if(isRangeOfSlices && is(typeof(copy(data, buf))))
+        {
+            size_t n = 0;
+            while(n < buf.length && !src.empty)
+            {
+                n += copy(data, buf[n .. $]);
+                if(data.empty)
+                {
+                    src.popFront;
+                    data = src.empty ? null : src.front;
+                }
+            }
+            return n;
+        }
+        else
+            static assert(false, "Incompatible read for type `", Buf, "` and source range `", R, "`");
+    }
+}
+
+/**
+ * This is an adapter that provides a way to "read" data from an input range.
+ * It can be used as an input source for any iopipe.
+ *
+ * This has a specialization for an input range that is a "range of slices"
+ * type, since we can utilize slice assignment for the copy.
+ */
+auto rangeInput(R)(R range) if (isInputRange!R)
+{
+    return RangeInput!R(range);
+}
+
+unittest
+{
+    import std.range;
+    import std.algorithm;
+    import iopipe.bufpipe;
+    import std.utf;
+
+
+    // make a big range of characters
+    {
+        auto bigRange = repeat(only(repeat('a', 10), repeat('b', 10))).joiner.joiner.take(100000);
+        auto inp = bigRange.rangeInput;
+        inp.read(cast(char[])[]);
```

Comment:
why is this done?

Reply:
I want to make sure a `read` compiles. When I didn't do this, and there was an error in the function, it instead failed the template constraints on `bufd`, and complained that I didn't pass in something properly, which is not a helpful error.

This makes sure if there is any error in the `read` function, it shows up as an actual error, instead of there. Try changing something in the function so it doesn't compile, and take out this line, you will see what I mean.

It should also be supported to read an empty buffer (though this shouldn't happen in normal code).